File: hrms/hr/page/paid_time_off/paid_time_off.py
# Copyright (c) 2026, Staff Pro BPO and Contributors
# License: GNU General Public License v3. See license.txt

import frappe
from frappe import _
from frappe.utils import flt, formatdate, get_year_ending, get_year_start, getdate

from hrms.hr.doctype.leave_application.leave_application import get_employee_leave_approver
from hrms.hr.report.employee_leave_balance.employee_leave_balance import get_data as get_leave_balance_data
from hrms.hr.utils import get_leave_period


HR_ROLES = ("HR Manager", "HR User", "System Manager", "Administrator")


def _ensure_hr():
	frappe.only_for(list(HR_ROLES))


def _active_employees() -> list[dict]:
	today = getdate()
	rows = frappe.get_list(
		"Employee",
		fields=["name", "employee_name", "company", "department", "image", "date_of_joining", "relieving_date"],
		filters={"status": "Active"},
		order_by="employee_name",
	)
	active = []
	for row in rows:
		if row.date_of_joining and getdate(row.date_of_joining) > today:
			continue
		if row.relieving_date and getdate(row.relieving_date) < today:
			continue
		active.append(
			{
				"name": row.name,
				"employee_name": row.employee_name or row.name,
				"company": row.company,
				"department": row.department,
				"image": row.image,
			}
		)
	return active


def _leave_types() -> list[dict]:
	return frappe.get_all(
		"Leave Type",
		fields=["name", "leave_type_name", "is_lwp"],
		order_by="leave_type_name",
	)


def _balance_period(company: str | None = None) -> tuple:
	today = getdate()
	if company:
		periods = get_leave_period(today, today, company) or []
		if periods:
			return getdate(periods[0].from_date), getdate(periods[0].to_date)
	return get_year_start(today), get_year_ending(today)


def _absence_status_class(status: str | None) -> str:
	status = str(status or "")
	if status == "Approved":
		return "approved"
	if status == "Open":
		return "open"
	if status == "Rejected":
		return "rejected"
	return "cancelled"


def _absence_status_label(status: str | None) -> str:
	status = str(status or "")
	if status == "Approved":
		return _("Approved")
	if status == "Open":
		return _("Pending")
	return status or _("—")


def _absence_date_label(from_date, to_date) -> str:
	if not from_date:
		return ""
	if from_date == to_date:
		return formatdate(from_date)
	return f"{formatdate(from_date)} – {formatdate(to_date)}"


def _employee_company(employee: str | None) -> str | None:
	if employee:
		return frappe.db.get_value("Employee", employee, "company")
	return frappe.defaults.get_user_default("Company")


def _balance_rows(employee: str | None = None) -> list[dict]:
	company = _employee_company(employee)
	from_date, to_date = _balance_period(company)
	filters = frappe._dict(
		{
			"from_date": from_date,
			"to_date": to_date,
			"consolidate_leave_types": 0,
			"employee_status": "Active",
		}
	)
	if company:
		filters.company = company
	if employee:
		filters.employee = employee

	rows = []
	for row in get_leave_balance_data(filters) or []:
		if not row.get("employee"):
			continue
		rows.append(
			{
				"leave_type": row.get("leave_type"),
				"employee": row.get("employee"),
				"employee_name": row.get("employee_name"),
				"opening_balance": flt(row.get("opening_balance")),
				"leaves_allocated": flt(row.get("leaves_allocated")),
				"leaves_taken": flt(row.get("leaves_taken")),
				"leaves_expired": flt(row.get("leaves_expired")),
				"closing_balance": flt(row.get("closing_balance")),
			}
		)
	return rows


def _absence_rows(employee: str | None = None, company: str | None = None) -> list[dict]:
	company = company or _employee_company(employee)
	if not company:
		return []

	today = getdate()
	LeaveApplication = frappe.qb.DocType("Leave Application")
	query = (
		frappe.qb.from_(LeaveApplication)
		.select(
			LeaveApplication.name,
			LeaveApplication.employee,
			LeaveApplication.employee_name,
			LeaveApplication.leave_type,
			LeaveApplication.from_date,
			LeaveApplication.to_date,
			LeaveApplication.description,
			LeaveApplication.status,
		)
		.where(LeaveApplication.company == company)
		.where(LeaveApplication.to_date >= today)
		.where(LeaveApplication.docstatus != 2)
		.where(LeaveApplication.status.isin(["Open", "Approved"]))
		.orderby(LeaveApplication.from_date)
		.orderby(LeaveApplication.employee_name)
	)
	if employee:
		query = query.where(LeaveApplication.employee == employee)

	rows = []
	for rec in query.run(as_dict=True):
		rows.append(
			{
				"leave_application": rec.name,
				"employee": rec.employee,
				"employee_name": rec.employee_name,
				"leave_type": rec.leave_type,
				"from_date": rec.from_date,
				"to_date": rec.to_date,
				"date_label": _absence_date_label(rec.from_date, rec.to_date),
				"description": rec.description or "",
				"status": rec.status,
				"status_label": _absence_status_label(rec.status),
				"status_class": _absence_status_class(rec.status),
			}
		)
	return rows


@frappe.whitelist()
def get_page_context(employee: str | None = None) -> dict:
	_ensure_hr()
	if employee:
		frappe.has_permission("Employee", "read", employee, throw=True)

	company = _employee_company(employee)
	from_date, to_date = _balance_period(company)
	return {
		"employees": _active_employees(),
		"leave_types": _leave_types(),
		"balances": _balance_rows(employee),
		"absences": _absence_rows(employee, company),
		"from_date": from_date,
		"to_date": to_date,
		"period_label": f"{formatdate(from_date)} – {formatdate(to_date)}",
	}


@frappe.whitelist()
def book_time_off(employee: str, leave_type: str, from_date: str, to_date: str | None = None) -> dict:
	_ensure_hr()
	frappe.has_permission("Leave Application", "create", throw=True)

	from_date = getdate(from_date)
	to_date = getdate(to_date or from_date)
	if to_date < from_date:
		frappe.throw(_("End date cannot be before start date."))

	approver = get_employee_leave_approver(employee) or frappe.session.user
	leave = frappe.get_doc(
		{
			"doctype": "Leave Application",
			"employee": employee,
			"leave_type": leave_type,
			"from_date": from_date,
			"to_date": to_date,
			"leave_approver": approver,
			"status": "Approved",
			"description": leave_type,
		}
	)
	leave.insert()
	if leave.status != "Approved":
		leave.status = "Approved"
	if leave.docstatus == 0:
		leave.submit()

	return {
		"name": leave.name,
		"message": _("Time off booked for {0}").format(leave.employee_name or employee),
	}


@frappe.whitelist()
def allocate_pto(employee: str, leave_type: str, days: float | str = 10) -> dict:
	_ensure_hr()
	frappe.has_permission("Leave Allocation", "create", throw=True)

	days = flt(days)
	if days <= 0:
		frappe.throw(_("Enter how many PTO days to allocate."))

	if frappe.db.get_value("Leave Type", leave_type, "is_lwp"):
		frappe.throw(_("{0} is unpaid and cannot be allocated.").format(leave_type))

	company = frappe.db.get_value("Employee", employee, "company")
	from_date, to_date = _balance_period(company)

	existing = frappe.db.get_value(
		"Leave Allocation",
		{
			"employee": employee,
			"leave_type": leave_type,
			"docstatus": 1,
			"from_date": ("<=", to_date),
			"to_date": (">=", from_date),
		},
		"name",
	)
	if existing:
		allocation = frappe.get_doc("Leave Allocation", existing)
		frappe.has_permission("Leave Allocation", "write", allocation, throw=True)
		allocation.new_leaves_allocated = flt(allocation.new_leaves_allocated) + days
		allocation.save()
		return {
			"name": allocation.name,
			"message": _("Added {0} day(s) of {1}").format(frappe.format_value(days, {"fieldtype": "Float"}), leave_type),
		}

	allocation = frappe.get_doc(
		{
			"doctype": "Leave Allocation",
			"employee": employee,
			"leave_type": leave_type,
			"from_date": from_date,
			"to_date": to_date,
			"new_leaves_allocated": days,
		}
	)
	allocation.insert()
	allocation.submit()
	return {
		"name": allocation.name,
		"message": _("Allocated {0} day(s) of {1}").format(frappe.format_value(days, {"fieldtype": "Float"}), leave_type),
	}

File: hrms/hr/page/paid_time_off/paid_time_off.js
frappe.provide("hrms");

frappe.pages["paid-time-off"].on_page_load = function (wrapper) {
	const page = frappe.ui.make_app_page({
		parent: wrapper,
		title: __("Paid Time Off"),
		single_column: true,
	});
	frappe.breadcrumbs.add("HR");
	hrms.paid_time_off.make(page);
};

frappe.pages["paid-time-off"].on_page_show = function () {
	hrms.paid_time_off.refresh();
};

hrms.paid_time_off = {
	page: null,
	$body: null,
	payload: null,
	employee: "",
	leave_type: "",
	busy: false,

	make(page) {
		this.page = page;
		const $existing = page.main.find(".sp-pto-page");
		if ($existing.length) {
			this.$body = $existing;
			this.refresh();
			return;
		}
		this.$body = $('<div class="sp-pto-page"></div>').appendTo(page.main);
		this.render_shell();
		this.bind();
		this.refresh();
	},

	escape(value) {
		return frappe.utils.escape_html(value == null ? "" : String(value));
	},

	today() {
		return frappe.datetime.get_today();
	},

	render_shell() {
		const today = this.today();
		this.$body.html(`
			<section class="sp-dash-panel sp-pto-form" aria-label="${this.escape(__("Book Time Off"))}">
				<div class="sp-dash-panel__head">
					<h2 class="sp-dash-panel__title">${this.escape(__("Paid Time Off"))}</h2>
				</div>
				<p class="sp-pto-form__hint">${this.escape(__("Pick an agent, the days they will be out, and the reason. Their PTO balance updates here and on Upcoming Absences."))}</p>
				<div class="sp-pto-form__grid">
					<label class="sp-pto-field">
						<span>${this.escape(__("Agent"))}</span>
						<select class="sp-pto-agent" aria-label="${this.escape(__("Agent"))}">
							<option value="">${this.escape(__("Select agent"))}</option>
						</select>
					</label>
					<label class="sp-pto-field">
						<span>${this.escape(__("From"))}</span>
						<input type="date" class="sp-pto-from" value="${this.escape(today)}" />
					</label>
					<label class="sp-pto-field">
						<span>${this.escape(__("To"))}</span>
						<input type="date" class="sp-pto-to" value="${this.escape(today)}" />
					</label>
					<label class="sp-pto-field">
						<span>${this.escape(__("Reason"))}</span>
						<select class="sp-pto-reason" aria-label="${this.escape(__("Reason"))}">
							<option value="">${this.escape(__("Vacation, sick leave…"))}</option>
						</select>
					</label>
					<label class="sp-pto-field sp-pto-field--days">
						<span>${this.escape(__("Days to add"))}</span>
						<input type="number" class="sp-pto-days" min="0.5" step="0.5" value="10" />
					</label>
				</div>
				<div class="sp-pto-form__actions">
					<button type="button" class="sp-pto-allocate">${this.escape(__("Allocate PTO"))}</button>
					<button type="button" class="sp-pto-book">${this.escape(__("Book Time Off"))}</button>
				</div>
			</section>
			<section class="sp-dash-panel sp-pto-balance" aria-label="${this.escape(__("PTO Balance"))}">
				<div class="sp-dash-panel__head">
					<h2 class="sp-dash-panel__title">${this.escape(__("PTO Balance"))}</h2>
					<p class="sp-pto-period"></p>
				</div>
				<div class="sp-pto-table-wrap">
					<div class="sp-pto-table-head">
						<span>${this.escape(__("Leave Type"))}</span>
						<span>${this.escape(__("Agent"))}</span>
						<span>${this.escape(__("Opening"))}</span>
						<span>${this.escape(__("Allocated"))}</span>
						<span>${this.escape(__("Taken"))}</span>
						<span>${this.escape(__("Expired"))}</span>
						<span>${this.escape(__("Closing"))}</span>
					</div>
					<div class="sp-pto-balance-list"></div>
				</div>
			</section>
			<section class="sp-dash-panel sp-pto-absences" aria-label="${this.escape(__("Upcoming Absences"))}">
				<div class="sp-dash-panel__head">
					<h2 class="sp-dash-panel__title">${this.escape(__("Upcoming Absences"))}</h2>
				</div>
				<div class="sp-pto-table-wrap">
					<div class="sp-pto-absence-head">
						<span>${this.escape(__("Agent"))}</span>
						<span>${this.escape(__("Absence Type"))}</span>
						<span>${this.escape(__("Dates"))}</span>
						<span>${this.escape(__("Status"))}</span>
					</div>
					<div class="sp-pto-absence-list"></div>
				</div>
			</section>
		`);
	},

	bind() {
		const me = this;
		this.$body.on("change", ".sp-pto-agent", function () {
			me.employee = $(this).val() || "";
			me.refresh();
		});
		this.$body.on("change", ".sp-pto-reason", function () {
			me.leave_type = $(this).val() || "";
		});
		this.$body.on("change", ".sp-pto-from", function () {
			const from = $(this).val();
			const $to = me.$body.find(".sp-pto-to");
			if (from && (!$to.val() || $to.val() < from)) {
				$to.val(from);
			}
		});
		this.$body.on("click", ".sp-pto-book", () => me.book());
		this.$body.on("click", ".sp-pto-allocate", () => me.allocate());
		this.$body.on("click", ".sp-pto-balance-row, .sp-pto-absence-row", function () {
			const leave = $(this).data("leave");
			const employee = $(this).data("employee");
			if (leave) {
				frappe.set_route("Form", "Leave Application", leave);
				return;
			}
			if (employee) {
				frappe.set_route("Form", "Employee", employee);
			}
		});
	},

	form_values() {
		return {
			employee: this.$body.find(".sp-pto-agent").val() || this.employee,
			leave_type: this.$body.find(".sp-pto-reason").val() || this.leave_type,
			from_date: this.$body.find(".sp-pto-from").val(),
			to_date: this.$body.find(".sp-pto-to").val(),
			days: this.$body.find(".sp-pto-days").val(),
		};
	},

	require_form(need_days = false) {
		const values = this.form_values();
		if (!values.employee) {
			frappe.msgprint(__("Select an agent."));
			return null;
		}
		if (!values.leave_type) {
			frappe.msgprint(__("Select a reason."));
			return null;
		}
		if (need_days && (!values.days || Number(values.days) <= 0)) {
			frappe.msgprint(__("Enter how many PTO days to allocate."));
			return null;
		}
		return values;
	},

	set_busy(busy) {
		this.busy = busy;
		this.$body.find(".sp-pto-book, .sp-pto-allocate").prop("disabled", busy);
	},

	book() {
		if (this.busy) return;
		const values = this.require_form();
		if (!values) return;
		this.set_busy(true);
		frappe.call({
			method: "hrms.hr.page.paid_time_off.paid_time_off.book_time_off",
			args: {
				employee: values.employee,
				leave_type: values.leave_type,
				from_date: values.from_date,
				to_date: values.to_date || values.from_date,
			},
			callback: (r) => {
				this.set_busy(false);
				frappe.show_alert({ message: r.message?.message || __("Time off booked"), indicator: "green" });
				this.refresh();
			},
			error: () => this.set_busy(false),
		});
	},

	allocate() {
		if (this.busy) return;
		const values = this.require_form(true);
		if (!values) return;
		this.set_busy(true);
		frappe.call({
			method: "hrms.hr.page.paid_time_off.paid_time_off.allocate_pto",
			args: {
				employee: values.employee,
				leave_type: values.leave_type,
				days: values.days,
			},
			callback: (r) => {
				this.set_busy(false);
				frappe.show_alert({ message: r.message?.message || __("PTO allocated"), indicator: "green" });
				this.refresh();
			},
			error: () => this.set_busy(false),
		});
	},

	refresh() {
		if (!this.$body) return;
		this.employee = this.$body.find(".sp-pto-agent").val() || this.employee;
		this.leave_type = this.$body.find(".sp-pto-reason").val() || this.leave_type;
		this.$body.find(".sp-pto-balance-list, .sp-pto-absence-list").addClass("is-loading");
		frappe.call({
			method: "hrms.hr.page.paid_time_off.paid_time_off.get_page_context",
			args: { employee: this.employee || undefined },
			callback: (r) => {
				this.payload = r.message || {};
				this.paint();
			},
			error: () => {
				this.$body.find(".sp-pto-balance-list, .sp-pto-absence-list").removeClass("is-loading");
			},
		});
	},

	paint() {
		const data = this.payload || {};
		this.fill_select(
			this.$body.find(".sp-pto-agent"),
			(data.employees || []).map((row) => ({
				value: row.name,
				label: row.employee_name || row.name,
			})),
			this.employee,
			__("Select agent"),
		);
		this.fill_select(
			this.$body.find(".sp-pto-reason"),
			(data.leave_types || []).map((row) => ({
				value: row.name,
				label: row.leave_type_name || row.name,
			})),
			this.leave_type,
			__("Vacation, sick leave…"),
		);
		this.$body.find(".sp-pto-period").text(data.period_label || "");
		this.render_balances(data.balances || []);
		this.render_absences(data.absences || []);
	},

	fill_select($select, options, value, placeholder) {
		const current = value || $select.val() || "";
		const html = [`<option value="">${this.escape(placeholder)}</option>`]
			.concat(
				options.map(
					(opt) =>
						`<option value="${this.escape(opt.value)}"${opt.value === current ? " selected" : ""}>${this.escape(opt.label)}</option>`,
				),
			)
			.join("");
		$select.html(html);
		if (current && options.some((opt) => opt.value === current)) {
			$select.val(current);
		}
	},

	number(value) {
		const amount = Number(value || 0);
		if (!Number.isFinite(amount)) return "0";
		return amount.toLocaleString(undefined, { maximumFractionDigits: 2 });
	},

	empty_html(message) {
		return `<div class="sp-pto-empty"><p>${this.escape(message)}</p></div>`;
	},

	render_balances(rows) {
		const $list = this.$body.find(".sp-pto-balance-list").removeClass("is-loading");
		if (!rows.length) {
			$list.html(this.empty_html(__("No PTO balances yet. Allocate PTO to get started.")));
			return;
		}
		$list.html(
			rows
				.map(
					(row) => `
				<button type="button" class="sp-pto-balance-row" data-employee="${this.escape(row.employee || "")}">
					<span>${this.escape(row.leave_type || "—")}</span>
					<span>${this.escape(row.employee_name || row.employee || "—")}</span>
					<span>${this.escape(this.number(row.opening_balance))}</span>
					<span>${this.escape(this.number(row.leaves_allocated))}</span>
					<span>${this.escape(this.number(row.leaves_taken))}</span>
					<span>${this.escape(this.number(row.leaves_expired))}</span>
					<span>${this.escape(this.number(row.closing_balance))}</span>
				</button>`,
				)
				.join(""),
		);
	},

	render_absences(rows) {
		const $list = this.$body.find(".sp-pto-absence-list").removeClass("is-loading");
		if (!rows.length) {
			$list.html(this.empty_html(__("No upcoming time off for this selection.")));
			return;
		}
		$list.html(
			rows
				.map(
					(row) => `
				<button type="button" class="sp-pto-absence-row" data-employee="${this.escape(row.employee || "")}" data-leave="${this.escape(row.leave_application || "")}">
					<span>${this.escape(row.employee_name || row.employee || "—")}</span>
					<span>${this.escape(row.leave_type || "—")}</span>
					<span>${this.escape(row.date_label || "")}</span>
					<span class="sp-payroll__status sp-payroll__status--${this.escape(row.status_class || "open")}">
						<span class="sp-payroll__status-dot" aria-hidden="true"></span>
						<span class="sp-payroll__status-label">${this.escape(row.status_label || row.status || "")}</span>
					</span>
				</button>`,
				)
				.join(""),
		);
	},
};
